### ISSUE TITLE
[FEATURE] Add bare-metal container and MCM support

### ISSUE THREAD

**Issue by andrewbabbittdev** [harvester/harvester#2679]
*Sorry if an existing ticket exists for this, hard to find anything similar with search*

**Is your feature request related to a problem? Please describe.**
Currently, the [rancher integration](https://docs.harvesterhci.io/v1.0/rancher/rancher-integration/) requires you to either have rancher deployed elsewhere, as a VM inside of harvester using docker (not recommended), or deployed in a cluster under harvester which you would have to manually deploy.

This is not ideal for many edge and lab oriented scenarios where you may not be running an external rancher instance. Dedicating additional resources to a cluster just for running rancher is non ideal in more resource constrained environments.

**Describe the solution you'd like**
[Rancher is already embedded in harvester](https://docs.harvesterhci.io/v1.0/troubleshooting/harvester/#access-embedded-rancher), it would be nice if cluster oriented features could be exposed to make it easy to deploy rke and k3s clusters directly through the harvester UI.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by yasker** [comment]
Depend on #2769 .

Also, the reason we don't support using embedded Rancher is it's very hard for the Harvester release to bind with the Rancher release. For Harvester, only a small subset feature of Rancher is used, e.g. managing node provisioning (RKE2) and managing upgrades process.

If we're going to support embedded Rancher, we would have to release Harvester every time when Rancher has a new release, thus resulting in the binding of the Harvester and the Rancher versions. It would be very challenging for the engineering team.

But we acknowledge there is a need to have a built-in Rancher. Currently, we're considering achieving it via #2769, like via creating one-click deployment for the Rancher HA cluster in Harvester. The embedded Rancher will still be there but it would still be reserved for debugging purposes only.

**Comment by davidspek** [comment]
For the time being, you can have a look at https://github.com/rancher/quickstart/pull/218.

**Comment by ibrokethecloud** [comment]
# Test plan:

The test plan is split into two parts. 

1. Install vcluster addon
2. Use vcluster deployed rancher to manage container workloads

If a user already has an external rancher cluster, managing harvester then vcluster addon is not needed, and we can directly validate container workload view. To leverage container workload view the minimal rancher version needs to be v2.7.5

## vcluster addon:
* an experimental `rancher-vcluster` addon is available at: https://github.com/harvester/experimental-addons/tree/main/rancher-vcluster
* to get start the user needs to deploy the yaml manifest to the harvester cluster via kubectl
* once kubeconfig is available simply run: `kubectl apply -f https://raw.githubusercontent.com/harvester/experimental-addons/main/rancher-vcluster/rancher-vcluster.yaml` 
![image](https://github.com/harvester/harvester/assets/19664014/11e0a6ad-5791-4b16-8363-f41bc4968863)
* the user should be able to see the `rancher-vcluster` addon 
![image](https://github.com/harvester/harvester/assets/19664014/a3a40bac-fbbd-44e5-8400-35a0192375ea)
* users can enable this addon by editing the config and providing a valid `Hostname` and `Bootstrap Password`
![image](https://github.com/harvester/harvester/assets/19664014/0df4ea4b-fe64-4a86-9526-fea860f32037)
* once the addon is deployed successfully, it can take a few mins for the rancher endpoint to be accessible
![image](https://github.com/harvester/harvester/assets/19664014/a93953c9-5050-40c5-96c9-333fa207d8e8)
* once rancher is online please follow the usual process to confirm settings
![image](https://github.com/harvester/harvester/assets/19664014/0dad13c5-6c9d-473e-90df-4bb66288ff5c)

This is a fully functional rancher, which can now be used to import the existing harvester cluster it is deployed on and provide full MCM capability

![image](https://github.com/harvester/harvester/assets/19664014/51c52e86-fe87-4004-a12f-75b6971ccc91)

Please refer to steps detailed in [official docs](https://docs.harvesterhci.io/v1.1/rancher/virtualization-management) for importing the cluster.

## container workload view
This assumes that user has imported harvester 1.2.0-rc4 or higher into a rancher v2.7.5 cluster

* once the harvester cluster is registered with rancher the users can click on `Manage` to access the explorer 
![image](https://github.com/harvester/harvester/assets/19664014/b7bb4d18-917c-4796-8454-a2c1b3e81984)

![image](https://github.com/harvester/harvester/assets/19664014/8f9d073a-f925-4424-972a-c3b370e253ca)

* once inside the `explore` view, the user should see the familiar rancher container workload management features
![image](https://github.com/harvester/harvester/assets/19664014/b136db7c-668b-4510-a010-843963354c65)

* users should only see the `Partners` chart repository, as the other repos are hidden to avoid deploying some common charts / integrations that are already packaged with harvester like monitoring and logging addons. Users can still creat and define their own chart repos.
![image](https://github.com/harvester/harvester/assets/19664014/20428c0b-475e-4e39-999e-d18fcb781ad5)

**Comment by lanfon72** [comment]
@guangbochen, @WuJun2016, I think UI should integration the addon before we start test it?
and the [HEP](https://github.com/harvester/harvester/pull/3673) would need to be update accordingly.

**Comment by guangbochen** [comment]
sure, and @lanfon72 we will update the HEP shortly.
For the UI integration, it is the intention not to add this experimental addon by default, but we will need a proper doc for this feature cc @ibrokethecloud.

**Comment by guangbochen** [comment]
link related issue on rancher https://github.com/rancher/rancher/issues/42200

**Comment by lanfon72** [comment]
@guangbochen, the [HEP](https://github.com/harvester/harvester/pull/3673) been closed by bot, will we initial another PR for it? and I can't find any reference to the MCM feature is asking the cluster MUST have FQDN or not.
![image](https://github.com/harvester/harvester/assets/5169694/ed60230a-81fd-44a9-82d5-73ad76c43150)

**Comment by guangbochen** [comment]
Yes, we will add a new HEP to reflect the latest design, for the FQDN error check if u have input the valid domain URL, e.g., you can use `rancher.$vip.sslip.io` if you don't have any custom DNS server.

**Comment by lanfon72** [comment]
> Yes, we add a new HEP to reflect the latest design, for the FQDN error check if u have input the valid domain URL, e.g., you can use `rancher.$vip.sslip.io` if you don't have any custom DNS server.

so this should be a bug right? :thinking: 
![image](https://github.com/harvester/harvester/assets/5169694/c2db421f-9013-4556-b6ac-e86755503ad2)

**Comment by lanfon72** [comment]
> > Yes, we add a new HEP to reflect the latest design, for the FQDN error check if u have input the valid domain URL, e.g., you can use `rancher.$vip.sslip.io` if you don't have any custom DNS server.
> 
> 
> 
> so this should be a bug right? :thinking: 
> 
> ![image](https://github.com/harvester/harvester/assets/5169694/c2db421f-9013-4556-b6ac-e86755503ad2)
> 
> 

Thanks @irishgordo 's help, it should be a UI bug.

I will test this feature first but still waiting the HEP to explain its scope and design.

**Comment by lanfon72** [comment]
add link to https://github.com/rancher/dashboard/issues/8800

**Comment by lanfon72** [comment]
Waiting the HEP to confirm this features' goals and non-goals.

**Comment by lanfon72** [comment]
According to [HEP](https://github.com/harvester/harvester/pull/4383):
> At present the installed `rancher` leverages a self signed certificate, and this can be extended in the future to allow users to brign their own certificates.

Following items will not be included in the test:
- [ ] Specific domain name for customized **hostname**
- [ ] SSL certificate for VCluster
- [ ] Air-gapped environment


Test Information
----
* Environment: **baremetal DL360G9 3 nodes**
* Harvester Version: **v1.2.0-rc5**
* **ui-source** Option: **Auto**

### Verify Steps
1. Install Harvester with any nodes
2. Login to console, apply _experimental addon: vcluster_ by executing `kubectl apply -f https://raw.githubusercontent.com/harvester/experimental-addons/main/rancher-vcluster/rancher-vcluster.yaml`
3. Login to dashboard, navigate to _Advanced/Addons_ then edit **rancher-vcluster**
4. Enable with customized `hostname` and `password`
5. **VCluster** should be deployed successfully
6. User should able to access **VCluster** by the customized `hostname`

### Regression Testing on Rancher of VCluster
  - ~~Local Rancher~~
  - RKE2
    - [x] create/upgrade v1.25.12+rke2r1 -> v1.26.7+rke2r1
    - [x] scale up/down 3 -> 4 -> 3 (nodes)
  - RKE1
    - [x] create/upgrade v1.25.12-rancher1-1 -> v1.26.7-rancher1-1
    - [x] scle up/down 3 -> 4 -> 3
  - Virtualization Management
    - [x] import parent
    - [x] import external and operations
    - [x] RBAC
    - [ ] Explore (Rancher): functionality are aligned to the *embedded* Rancher in Harvester, but UI aren't.

**Comment by lanfon72** [comment]
For the Explore (rancher) UI, tested and describe in https://github.com/rancher/dashboard/issues/8800#issuecomment-1666934150

@WuJun2016, I will put this back to implement, the **Bootstrap Password** field should be _required_, or we should hint the default password.
![image](https://github.com/harvester/harvester/assets/5169694/312f00b8-a21f-4e17-b612-195970febf1d)

**Comment by lanfon72** [comment]
> For the Explore (rancher) UI, tested and describe in [rancher/dashboard#8800 (comment)](https://github.com/rancher/dashboard/issues/8800#issuecomment-1666934150)
> 
> @WuJun2016, I will put this back to implement, the **Bootstrap Password** field should be _required_, or we should hint the default password. ![image](https://user-images.githubusercontent.com/5169694/259148972-312f00b8-a21f-4e17-b612-195970febf1d.png)

Tested on `v1.2.0-rc5` with **ui-source: External**, this change has been applied.

We would still wait the https://github.com/harvester/harvester/pull/4383#discussion_r1289510761 be included in.

**Comment by lanfon72** [comment]
As tested result, re-enable Vcluster will reuse the existing PVC and change **hostname** or **Bootstrap Password** is not working.  I think we would need (at least) to note this behavior in document.

And there have 2 things need to be documented or changed:
1. **BootstrapPassword** is plain text when **Edit as YAML**, it might not be a good practice.
2. **rancherVersion** is described[^1] to change/upgrade the rancher, but it is not working.

![image](https://github.com/harvester/harvester/assets/5169694/50b87ff5-2ac7-437e-8968-e3fcc54b27a0)

[^1]: https://github.com/harvester/experimental-addons/tree/main/rancher-vcluster#rancher-vcluster

Test Information
----
* Environment: **qemu/KVM 2 nodes**
* Harvester Version: **v1.2-3dc52b6b-head**
* **ui-source** Option: **Auto**

### Verify Steps
- Same as https://github.com/harvester/harvester/issues/2679#issuecomment-1669828538

**Comment by WuJun2016** [comment]
> rancherVersion is described[1](https://github.com/harvester/harvester/issues/2679#user-content-fn-1-d6ff445775746604274fd6eae96a4e50) to change/upgrade the rancher, but it is not working.

@lanfon72   Upgrading is not recommended from the UI at the moment, it requires filtering incompatible versions, so we have disabled the edit rancherVersion feature.

**Comment by lanfon72** [comment]
Verified this feature has been implemented.

After [this change](https://github.com/harvester/experimental-addons/pull/2), `rancherVersion` could be changed correctly; to use release candidates, we need to update `repo: https://releases.rancher.com/server-charts/stable/` to `repo: https://releases.rancher.com/server-charts/latest/` in the yaml file.

Other caveats will be improve in documents, so I will close this issue.

Test Information
----
* Environment: **Baremetal DL360G9 3 nodes**
* Harvester Version: **v1.2.0-rc5**
* **ui-source** Option: **External**

### Verify Steps
- Same as https://github.com/harvester/harvester/issues/2679#issuecomment-1669828538
